flang/lib/Lower/OpenMP/OpenMP.cpp

```diff
@@ -455,20 +489,23 @@ static void genNestedEvaluations(lower::AbstractConverter &converter,
     converter.genEval(e);
 }
 
-static bool
-mustEvalTeamsThreadsOutsideTarget(const lower::pft::Evaluation &eval,
-                                  mlir::omp::TargetOp targetOp) {
-  if (!targetOp)
-    return false;
-
-  auto offloadModOp = llvm::cast<mlir::omp::OffloadModuleInterface>(
-      *targetOp->getParentOfType<mlir::ModuleOp>());
-  if (offloadModOp.getIsTargetDevice())
+static bool mustEvalTeamsOutsideTarget(const lower::pft::Evaluation &eval,
+                                       mlir::omp::TargetOp targetOp) {
+  if (!isHostTarget(targetOp))
     return false;
 
   llvm::omp::Directive dir =
       extractOmpDirective(eval.get<parser::OpenMPConstruct>());
-  return llvm::omp::allTargetSet.test(dir) || !evalHasSiblings(eval);
+  return llvm::omp::allTeamsSet.test(dir) &&
```

Comment:
From the function name, it seems like `llvm::omp::allTeamsSet.test(dir)` is pre-condition (i.e. we only invoke the function from where we handle teams-related directives). Should that be an `assert`?